keps/sig-auth/0034-20190123-harden-default-discovery-bindings.md

```diff
@@ -0,0 +1,91 @@
+
+## Proposal
+
+* Remove the `system:unauthenticated` subject group from the default [`system:discovery` and `system:basic-user` ClusterRoleBindings](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L526-L527)
+* Create a new default ClusterRole which prominently, explicitly documents its intent to freely disclose information (e.g. `system:public-info-viewer`) and contains the `GET /healthz`, `/version` and `/version/` PolicyRules.
+* Create a new default ClusterRoleBinding for the new ClusterRole, which grants the `system:(un)authenticated` subject groups access.
+
+### Risks and Mitigations
+
+#### Existing customization of `system:discovery`
+
+If a user upgrades a cluster which has modified the `system:discovery` ClusterRoleBinding, these changes could either be trampled or restricted endpoints could end up being re-exposed via the new `system:public-info-viewer` binding.
+
+During the API server's RBAC auto-reconciliation, if `system:discovery` exists and `system:public-info-viewer` does not (i.e. the state after an upgrade) we'll copy `system:discovery`'s subjects and reconciliation annotation value. This will preserve permissions as-is during API server upgrades, even with the addition of `system:public-info-viewer`, since existing customization will be preserved.
+
+#### Dependence on existing unauthenticated behavior
+
+Some use-cases might require the existing permissions to be preserved for unauthenticated calls
+
+In the release notes, we can include an easy 'escape hatch' to re-enable unauthenticated access to the APIs, e.g. `kubectl create clusterrolebinding public-discovery --clusterrole=system:discovery --group=system:unauthenticated`
+
+From [`system:basic-user`](https://github.com/kubernetes/kubernetes/blob/8b98e802eddb9f478ff7d991a2f72f60c165388a/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L209-L215):
+* create `selfsubjectaccessreviews`
+* create `selfsubjectrulesreviews`
+
+pro: unauthenticated users can't list actions available to them
+con: this doesn't actually forbid actions, and can create a potentially confusing mismatch for users which rely on these APIs as a sanity check
+
+From [`system:discovery`](https://github.com/kubernetes/kubernetes/blob/8b98e802eddb9f478ff7d991a2f72f60c165388a/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L198-L208):
+* get `/api`, `/api/*`
+* get `/apis`, `/apis/*`
+* get `/openapi`, `/openapi/*`
+
+pro: can't see potentially sensitive custom resource names/schemas by default
+con: consumers that need schema info (e.g. editors) must allow authentication to the cluster to fetch it (though, this is already the case with clusters which have disabled anonymous auth)
+
```

Comment:
Be sure to include graduation criteria from the [KEP template](https://github.com/kubernetes/enhancements/blob/master/keps/0000-kep-template.md), including any test plan and documentation updates required